plugins/search-backend-module-elasticsearch/src/engines/ElasticSearchSearchEngine.ts

```diff
@@ -0,0 +1,265 @@
+/*
+ * Copyright 2021 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  IndexableDocument,
+  SearchQuery,
+  SearchResultSet,
+  SearchEngine,
+} from '@backstage/search-common';
+import { Logger } from 'winston';
+import esb from 'elastic-builder';
+import { Client } from '@elastic/elasticsearch';
+import { Config } from '@backstage/config';
+import {
+  createAWSConnection,
+  awsGetCredentials,
+} from '@acuris/aws-es-connection';
+import { isEmpty, isNaN as nan, isNumber } from 'lodash';
+
+export type ConcreteElasticSearchQuery = {
```

Comment:
This pattern of `documentTypes` and `*Builder` is probably not necessary for ES; it was a workaround to the limitations of Lunr's query builder / syntax, but ES' query language is _very_ robust, and I think we can get away with just returning a static (but typed) query JSON object.

For simplicity, I might recommend tying this type directly to the type accepted by `this.elasticSearchClient.search()`, thus freeing up the `query()` method to more purely be "translate query -> perform query -> transform results"

Reply:
Yup, makes sense. I was wondering why a constructing function was returned for Lunr but apparently it's just a limitation of the engine. 

I think documentTypes will make sense in the return object though. This would be the only way to easily modify the search functionality without the need to recreate the whole implementation so having the possibility to return/omit/modify which document types (which for some usages can be considered some kind of filters) to include in the final search call is something that is probably good to support.